### ISSUE TITLE
Add support for Azure ML Datasets

### ISSUE THREAD

**Issue by marrrcin** [getindata/kedro-azureml#10]
Azure ML Studio provides has a `Dataset` abstraction which allows to register datasets with some data, validate it's schema etc.
It would be nice to have a integration with this piece of Azure ML's SDK in form of custom Kedro DataSet to allow storing and retrieving the data directly from Azure. This would be a next level of integration with AML service.

https://docs.microsoft.com/en-us/azure/machine-learning/how-to-create-register-data-assets?tabs=Python-SDK

**Comment by fdroessler** [comment]
Are you already actively working on this (or have a specific design in mind)? Otherwise this would be something interesting for me in a current project so I could have a look at it.

**Comment by marrrcin** [comment]
Thanks for your interest in the topic @fdroessler ! We would greatly appreciate contribution to this from your side, so go ahead!

As for the design, we wanted this to be implemented as a [custom Kedro DataSet](https://kedro.readthedocs.io/en/stable/extend_kedro/custom_datasets.html) (one or more, depending on underlying implementations in Azure SDK v2). In general, the goal is to allow users to use Azure ML DataSet(s) in the Kedro's Data Catalog like this:
```yaml
# catalog.yml
my_dataset:
    type: kedro_azureml.datasets.AzureDataSet  # just an example name
    < parameters related to Azure Dataset>
```

in both read and write (if supported). The implementation should allow to read/write this dataset no matter where you're running the pipeline - either on Azure ML Pipelines or locally.

**Comment by tomasvanpottelbergh** [comment]
I was looking into how we could implement such a custom Kedro DataSet, but it seems the AML SDK v2 doesn’t support downloading registered datasets locally. This means it would only be possible to use them in Azure ML jobs, which we would probably like to avoid. The workarounds I can see would be:
- Using the v1 SDK for Datasets, which allow local up- and downloading. This would be the easiest, but [not recommended](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-migrate-from-v1#can-i-use-v1-and-v2-together) by Microsoft.
- Using the v2 SDK for Data to do the local upload, and using its path pointer to download the data using e.g. the Azure Storage SDK. This leaves open the question of how to download from Datastore URIs like `azureml://datastores/<datastore_name>/paths/<folder>` (see https://docs.microsoft.com/en-us/azure/machine-learning/how-to-migrate-from-v1#can-i-use-v1-and-v2-together).

Do you have any thoughts on this?

**Comment by marrrcin** [comment]
From what I see here https://docs.microsoft.com/en-us/azure/machine-learning/how-to-read-write-data-v2?tabs=python - in order to read the data, it need's to be passed as an input to the command (in our case - an invocation of a single kedro node). Maybe we can leverage that somehow?

I agree however that it's only possible to use them in Jobs, which is disappointing 😔

We can either wait for Microsoft (maybe they will implement better API for datasets in v2) or just use v1. In my opinion, if it's impossible to load the data "manually" in v2, e.g. into notebook running in Azure, it's really inconvenient and seems like a downgrade in the API...

**Comment by tomasvanpottelbergh** [comment]
I explored the options a bit more in detail and using v2 in the way you describe looks the cleanest for now, even though it will only work in Jobs. We could then overwrite the Data Catalog in `conf/local` to make the pipelines run locally as well. 

Hopefully Microsoft adds the ability to download the data to the v2 of the SDK, I created https://github.com/Azure/azure-sdk-for-python/issues/26213 to request this.

**Comment by marrrcin** [comment]
Thanks for creating the issue there, I've "upvoted" it. Let's see what happens :)

**Comment by marrrcin** [comment]
@tomasvanpottelbergh seems like Azure has released fsspec for Azure datasets https://learn.microsoft.com/en-us/azure/machine-learning/how-to-access-data-interactive?tabs=adls - are you planning to integrate it in our plugin?

**Comment by tomasvanpottelbergh** [comment]
I have another way of using Azure ML datasets which I intend to contribute, but the fsspec approach might be worth a try. I think @fdroessler was working on an approach that might be more in line with fsspec?

**Comment by fdroessler** [comment]
Is it just me or is working with Data using AzureML Studio just messy? Probably just me.

I have two "working" prototypes (please consider them quick and dirty rather than perfectly thought through) and I'd be curious what others think. Both "wrap" standard kedro dataset types so should work with a variety of kedro supported datasets.

Some of the thoughts on the first version were to preserve a potential folder structure in the underlying storage as opposed to have a lot of "local uploads".

The first version that I created before `azureml` had fsspec support, basically used kedro native versioned datasets and e.g. blob storage plus registering it in azureml after upload. Here the "downside" is that we'll still need the storage account credentials. Also I am not sure if registering them in AML is strictly necessary?

```
class KedroAMLDataset(AbstractDataSet):
    def __init__(
        self,
        name: str = None,
        aml_version: str = None,
        description: str = None,
        dataset: Union[str, Type[AbstractDataSet], Dict[str, Any]] = None,
        credentials: Dict[str, Any] = None,
        load_args: Dict[str, Any] = None,
        fs_args: Dict[str, Any] = None,
    ):
        # pylint: disable=import-outside-toplevel
        from fsspec.utils import infer_storage_options  # for performance reasons

        self._path = dataset["filepath"]
        self._description = description
        super().__init__()

        self._protocol = infer_storage_options(self._path)["protocol"]

        dataset = dataset if isinstance(dataset, dict) else {"type": dataset}
        self._dataset_type, self._dataset_config = parse_dataset_definition(dataset)
        if VERSION_KEY not in self._dataset_config:
            self._logger.warning(
                f"'{self.__class__.__name__}' only works with versioning of the "
                f"underlying dataset. Versioning will be added by default in"
                f"the dataset definition."
            )
            self._dataset_config[VERSION_KEY] = True

        if credentials:
            if CREDENTIALS_KEY in self._dataset_config:
                self._logger.warning(
                    KEY_PROPAGATION_WARNING,
                    {"keys": CREDENTIALS_KEY, "target": "underlying dataset"},
                )
            else:
                self._dataset_config[CREDENTIALS_KEY] = deepcopy(credentials)

        self._credentials = deepcopy(credentials) or {}

        self._fs_args = deepcopy(fs_args) or {}
        if self._fs_args:
            if "fs_args" in self._dataset_config:
                self._logger.warning(
                    KEY_PROPAGATION_WARNING,
                    {"keys": "filesystem arguments", "target": "underlying dataset"},
                )
            else:
                self._dataset_config["fs_args"] = deepcopy(self._fs_args)

        self._load_args = deepcopy(load_args) or {}
        credential = DefaultAzureCredential()
        self._ml_client = MLClient.from_config(
            credential=credential, path="conf/config.json"
        )
        self._name = name
        self._aml_version = aml_version

    def _load(self) -> pd.DataFrame:
        if self._aml_version:
            filedataset_asset = self._ml_client.data.get(
                name=self._name, version=self._aml_version
            )
        else:
            filedataset_asset = self._ml_client.data.get(
                name=self._name, label="latest"
            )
        underlying_dataset_version = VERSION_REGEX.findall(filedataset_asset.path)
        if underlying_dataset_version:
            underlying_dataset_version = underlying_dataset_version[0]
        load_version = Version(
            underlying_dataset_version, self._dataset_config["version"].save
        )
        self._dataset_config["version"] = load_version

        dataset = self._dataset_type(**self._dataset_config)
        return dataset.load()

    def _save(self, df: pd.DataFrame) -> None:
        dataset = self._dataset_type(**self._dataset_config)
        azure_fp = f"{dataset._protocol}://{dataset._get_save_path()}"
        dataset.save(df)
        aml_dataset = Data(
            path=azure_fp,
            name=self._name,
            description=self._description,
            type=AssetTypes.URI_FILE,
        )
        self._ml_client.data.create_or_update(aml_dataset)

    def _exists(self) -> bool:
        path = self._get_load_path()
        return Path(path.as_posix()).exists()

    def _describe(self):
        return dict(
            version=self._aml_version, name=self._name, description=self._description
        )
```

With the respective `catalog.yaml` entry:

```
data_processing.preprocessed_shuttles:
  type: spaceflights.dataset2.KedroAMLDataset
  name: preprocessed_shuttles
  # aml_version: "1"
  description: "this is the test dataset"
  dataset:
    type: pandas.CSVDataSet
    filepath: abfss://kedro-test/preprocessed_shuttles.csv
    versioned: True
```

The second version uses fsspec for reading. So far I was unable to figure out how to select a datastore different than the default one for uploading data. Any ideas? Currently the "downside" here is that all datasets are "saved" locally before being uploaded as manual uploads to AML. It won't support versioned datasets from kedro at all at the moment.

```
class KedroAMLDataset_fsspec(AbstractDataSet):
    def __init__(
        self,
        name: str = None,
        aml_version: str = None,
        description: str = None,
        dataset: Union[str, Type[AbstractDataSet], Dict[str, Any]] = None,
        credentials: Dict[str, Any] = None,
        load_args: Dict[str, Any] = None,
        fs_args: Dict[str, Any] = None,
    ):
        # pylint: disable=import-outside-toplevel
        from fsspec.utils import infer_storage_options  # for performance reasons

        self._path = dataset["filepath"]
        self._description = description
        super().__init__()

        self._protocol = infer_storage_options(self._path)["protocol"]

        dataset = dataset if isinstance(dataset, dict) else {"type": dataset}
        self._dataset_type, self._dataset_config = parse_dataset_definition(dataset)

        if credentials:
            if CREDENTIALS_KEY in self._dataset_config:
                self._logger.warning(
                    KEY_PROPAGATION_WARNING,
                    {"keys": CREDENTIALS_KEY, "target": "underlying dataset"},
                )
            else:
                self._dataset_config[CREDENTIALS_KEY] = deepcopy(credentials)

        self._credentials = deepcopy(credentials) or {}

        self._fs_args = deepcopy(fs_args) or {}
        if self._fs_args:
            if "fs_args" in self._dataset_config:
                self._logger.warning(
                    KEY_PROPAGATION_WARNING,
                    {"keys": "filesystem arguments", "target": "underlying dataset"},
                )
            else:
                self._dataset_config["fs_args"] = deepcopy(self._fs_args)

        self._load_args = deepcopy(load_args) or {}
        credential = DefaultAzureCredential()
        self._ml_client = MLClient.from_config(
            credential=credential, path="conf/config.json"
        )
        self._name = name
        self._aml_version = aml_version

    def _load(self) -> pd.DataFrame:
        if self._aml_version:
            filedataset_asset = self._ml_client.data.get(
                name=self._name, version=self._aml_version
            )
        else:
            filedataset_asset = self._ml_client.data.get(
                name=self._name, label="latest"
            )
        ds_config = self._dataset_config
        ds_config['filepath'] = filedataset_asset.path
        dataset = self._dataset_type(**self._dataset_config)
        return dataset.load()

    def _save(self, df: pd.DataFrame) -> None:
        dataset = self._dataset_type(**self._dataset_config)
        save_path = dataset._get_save_path()
        # this could also just be a temporary folder later?
        dataset.save(df)
        aml_dataset = Data(
            path=save_path,
            name=self._name,
            description=self._description,
            type=AssetTypes.URI_FILE,
        )
        self._ml_client.data.create_or_update(aml_dataset)

    def _exists(self) -> bool:
        path = self._get_load_path()
        return Path(path.as_posix()).exists()

    def _describe(self):
        return dict(
            version=self._aml_version, name=self._name, description=self._description
        )
```

Corresponding `catalog.yaml`:

```
data_processing.preprocessed_companies:
  type: spaceflights.dataset2.KedroAMLDataset_fsspec
  name: preprocessed_companies
  # aml_version: "1"
  description: "this is the test dataset"
  dataset:
    type: pandas.CSVDataSet
    filepath: preprocessed_companies.csv
```

Anyway it feels like there should be more elegant solutions so I'll keep thinking about this over the next days.

**Comment by marrrcin** [comment]
There are a few things here that I would consider:
1. Is the fsspec implementing `azureml://` (https://learn.microsoft.com/en-us/azure/machine-learning/how-to-access-data-interactive?tabs=adls) good enough to use directly with existing Kedro datasets (such as CSVDataSet / ParquetDataSet etc)? If so, we should add the `azureml-fsspec` as a dependency and provide some documentation for that.
2. Tighter integration with AML datasets (registering etc) - your "nested" dataset solution looks good to me on the conceptual level, I'm not sure about compatibility there - what if someone uses dataset that e.g. does not have `filepath` attribute?

**Comment by eliorc** [comment]
Work in progress at #49

**Comment by marrrcin** [comment]
Closed via https://github.com/getindata/kedro-azureml/pull/49
Released in `0.4.0`
